File: week-5/longest-univalue-path/longest-univalue-path.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var longestUnivaluePath = function(root) {
    const res = [0];
    recurse(root, res);
    return res[0];
};

const recurse = (root, res) => {
    if (!root) return;
    
    const left = dfs(root.left, root.val);
    const right = dfs(root.right, root.val);
    
    res[0] = Math.max(res[0], left+right);
    
    recurse(root.left, res);
    recurse(root.right, res);
};

const dfs = (root, cur) => {
    if (!root || cur !== root.val) return 0;
    
    return 1 + Math.max(dfs(root.left, root.val), dfs(root.right, root.val));
}

File: week-5/longest-univalue-path/longest-univalue-path.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def longestUnivaluePath(self, root: TreeNode) -> int:
        res = [0]
        self.recurse(root, res)
        return res[0]
    
    def recurse(self, root, res):
        if not root:
            return
        left = self.dfs(root.left, root.val)
        right = self.dfs(root.right, root.val)
        res[0] = max(res[0], left+right)
        
        self.recurse(root.left, res)
        self.recurse(root.right, res)
    
    def dfs(self, root, cur):
        if not root or cur != root.val: return 0
        
        return 1+ max(self.dfs(root.left, root.val), self.dfs(root.right, root.val))
